src/cpu/flexcpu/simple_dataflow_cpu.cc

```diff
@@ -0,0 +1,158 @@
+        // the SimpleCPU implementations have.
+        threads.push_back(SDCPUThreadState(this, i, params->system,
+            params->workload[i], params->itb, params->dtb, params->isa[i]));
+        threadContests.push_back(threads[i].getThreadContext());
+    }
+    // TODO add any other constructor details
+}
+
+SimpleDataflowCPU::~SimpleDataflowCPU()
+{
+    //
+}
+
+void
+SimpleDataflowCPU::activateContext(ThreadID tid)
+{
+    DPRINTF(SimpleDataflowCPU, "ActivateContext %d\n", tid);
+
+    BaseCPU::activateContext(tid);
+    // TODO activate the thread
+
+    // TODO schedule the first fetch event.
+    // schedule(new EventFunctionWrapper([this]{ fetch(); },
+    //                                   name() + ".initial_fetch",
+    //                                   true),
+    //          curTick());
+}
+
+MasterPort&
+SimpleDataflowCPU::getDataPort() {
+    return _dataPort;
+}
+
+MasterPort&
+SimpleDataflowCPU::getInstPort() {
+    return _instPort;
+}
+
+void
+SimpleDataflowCPU::init()
+{
+    BaseCPU::init();
+
+    if (!params()->switched_out &&
+        system->getMemoryMode() != Enums::timing)
+    {
+        fatal("The Dataflow CPU requires the memory system to be in "
+            "'timing' mode.\n");
+    }
+
+    for (SDCPUThread& thread : threads) {
+        ThreadContext* tc = thread.getThreadContext();
+        tc->initMemProxies(tc);
+
+        // TODO if fullsystem, we need to check if we need to call initCPU on
+        // the ISA
+    }
+    // TODO add post-construction initialization code
+}
+
+void
+SimpleDataflowCPU::startup()
+{
+    BaseCPU::startup();
+
+    for (SDCPUThread& thread : threads) {
+        thread.startup();
+    }
+    // TODO add code which needs to be run during the first simulation event
+}
+
+void
+SimpleDataflowCPU::switchOut()
+{
+    // TODO flush speculative state.
+    // TODO clear any outstanding memory requests.
+
+    BaseCPU::switchOut();
+}
+
+void
+SimpleDataflowCPU::takeOverFrom(BaseCPU* cpu)
+{
+    // TODO may want to ensure clean speculative state prior to attempting?
+
+    BaseCPU::takeOverFrom(cpu); // BaseCPU::takeOverFrom() already handles the
+                                // calls to ThreadContext::takeOverFrom() via
+                                // the internal ThreadContext* table.
+
+    // TODO load any other state.
+}
+
+Counter
+SimpleDataflowCPU::totalInsts()
+{
+    return 0; // TODO actually get a count of total instructions
+}
+
+Counter
+SimpleDataflowCPU::totalOps()
+{
+    return 0; // TODO actually get a count of total operations
+}
+
+void
+SimpleDataflowCPU::wakeup(ThreadID tid)
+{
+    assert(tid < numThreads);
+
+    // Get the ThreadContext interface pointer from BaseCPU::getContext(), for
+    // flexibility in how the ThreadContext is retrieved (in case we change
+    // from using SimpleThread).
+    ThreadContext* tc = getContext(tid);
+    if (tc->status() == ThreadContext::Suspended) {
+        tc->activate();
+    }
+}
```

Comment:
No newline. I'll stop commenting on this, but it should be fixed in the rest of the code.